### PR TITLE
Reuse published vectors for files a refresh did not change

### PR DESCRIPTION
**Pull request by basnijholt**
Every semantic knowledge refresh re-embeds the whole corpus, however little
changed. A base that takes an hour to index takes an hour again to absorb a
one-line edit, and pays the provider for it.

## The defect

Publishing a candidate index deletes its checkpoint. The next refresh therefore
finds no checkpoint in `_open_candidate_run`, mints a fresh candidate collection
with `completed={}`, resets it, and starts from zero — even though the
collection it just published holds vectors for every file that did not change.

Reproduced on a three-file base with the recording embedder:

```
1st refresh: indexed 3 files
change one file
2nd refresh: indexed 3 files, embedded "content 0", "content 1", "content 2"
```

`test_refresh_after_publish_reuses_vectors_for_unchanged_files` fails on `main`
with `AssertionError: the whole corpus was embedded again / assert 3 == 1`.

The cost is O(corpus) per refresh where the delta is O(changed files), so it
grows without bound as a base grows, and it repeats on every poll interval.

## The mechanism

The published collection is already an index from source path to vectors. Every
chunk was written with the metadata `_index_file_locked` attaches:

```python
metadata = {
    _SOURCE_PATH_KEY: relative_path,
    _SOURCE_MTIME_NS_KEY: source_mtime_ns,
    _SOURCE_SIZE_KEY: source_size,
    _SOURCE_DIGEST_KEY: source_digest,
}
```

So a fresh candidate no longer starts empty. It copies those chunks **verbatim**
— ids, embeddings, documents and metadata — for every currently-managed path,
and records each copied path in the checkpoint with the signature the published
index stored for it. Only paths the copy did not cover are embedded.

Nothing new decides whether the copy was right. `_reconcile_candidate` already
runs immediately afterwards, compares every recorded signature against the file
on disk, and drops and re-indexes whatever moved. Reuse is an optimization
layered on the existing correctness check, not a second one.

This deliberately does not use a Git diff. The digest already in the metadata is
a stronger signal than a revision range, and reconciliation already enforces it,
so a diff would only avoid copying rows for the few files that changed — a
fraction of one refresh's local I/O, in exchange for a subprocess call, a
fallback path, and a Git-only restriction. Without it the reuse also covers
non-Git watched folders.

## When reuse is allowed

All of these must hold, or the candidate is built from zero exactly as before:

| Condition | Why |
|---|---|
| `IndexingSettings` are equal | Pins the chunker (`chunk_size`, `chunk_overlap`), the embedder identity (provider, model, host, dimensions) and every corpus filter (branch, LFS, hidden, include/exclude patterns and extensions). Identical bytes under identical settings produce identical chunks and identical vectors. |
| Persisted status is `complete` | Any other collection is one the candidate lifecycle treats as an abandoned candidate it may reclaim. Copying from one would read a collection nothing keeps alive. |
| A collection is named | Type narrowing for the copy call. Unreachable at runtime: `status_when_missing="indexing"` is only written when there is no prior state, so no collection is named. |
| The refresh is not forced | An operator forcing a rebuild does not trust the index. A forced refresh both suppresses the copy and discards any surviving candidate, because that candidate's claims are the same suspect vectors. |
| The copy succeeds | Any exception resets the candidate and returns nothing, so the refresh degrades to a full rebuild instead of failing. |

Per path, the claim is derived from the rows actually copied, never from the
corpus listing, so a path the published index cannot serve is never claimed and
gets embedded normally.

## Recovering an interrupted copy

The copy writes rows before it records them, so an interruption can leave the
candidate holding vectors its checkpoint does not account for. That matters
because reconciliation derives every deletion from the checkpoint: rows outside
it are invisible to it and would ride into the published index — including rows
for a path that has since left the corpus, which would then keep serving its old
content from search.

`asyncio.to_thread` cannot be cancelled, so this is not exotic: the copy thread
runs to completion while the awaiting refresh raises `CancelledError`, and
`KnowledgeRefreshScheduler.shutdown()` cancels in-flight refreshes on any
ordinary restart.

Rather than record that a copy is in progress, the candidate carries an
invariant: **its collection may only hold rows the checkpoint claims.**

The published-vector copy writes many paths before recording any, then records
every copied path in one write after the last row lands. Ordinary indexing also
has an unclaimed window, but records each file immediately after its rows land,
so that window is bounded by the in-flight file rather than the whole copied
corpus. In either case, no claims plus a non-empty collection means rows exist
outside checkpoint accounting, which one bounded query detects when the
candidate is opened. A candidate in that state is rebuilt and copied again
rather than resumed over.

Two things follow that a stored flag would not have given:

- There is nothing to set early or clear late, so the class of ordering bugs a
  marker invites does not exist. The same check also covers a crash *during* a
  rebuild, between the claims being dropped and the collection being emptied.
- No persisted format change, which is the one part of this that would have been
  expensive to undo.

Claims are **not** written per page, and that is load-bearing rather than
incidental: a file's chunks can span pages, so a partial claim would name a path
holding only some of its chunks, and `_candidate_paths_missing_vectors` confirms
a path from a single row — the candidate would publish a silently truncated
file. `test_a_copy_records_its_claims_only_after_every_row_has_landed` pins it,
because nothing else in the design would notice.

## The traps

**1. mtime restamping.** Copied rows carry the published mtime, which a checkout
routinely invalidates without changing content. Recording that signature is
correct precisely because `_reconcile_candidate` compares only size and digest to
decide whether vectors survive, and adopts a new stamp for byte-identical
content through `_restamp_candidate_paths` — one batched journal append, no
re-embedding. The test bumps every mtime and asserts the *embed* count, not the
file count.

**2. Row-bounded copy query.** Chroma binds one SQL variable per *returned* row
and SQLite caps a statement at 32,766, so nothing about the file or path count
bounds a copy query. Measured against the pinned ChromaDB 1.5.8, on a collection
of 40,000 rows:

| Query | Result |
|---|---|
| `get(include=[embeddings, documents, metadatas])`, unpaged | `too many SQL variables` |
| same, `limit=1000` | ok |
| same, `limit=20000` | ok |
| same, `limit=40000` | `too many SQL variables` |

The copy pages at `_PUBLISHED_VECTOR_COPY_PAGE_ROWS = 500`, ~40x under the
observed ceiling, which also bounds how much vector data is held at once (a
3072-dimension embedder is ~12 MB of float per page).

Reuse also *removes* pressure from the query #1707 had to teach itself to split:
copied paths are marked `verified`, because the copy already reported which
paths received rows, so they never reach `_candidate_paths_missing_vectors`.

**3. Zero-length files.** They legitimately produce no vectors, so no copy can
prove them indexed; they are not claimed, and are rescanned at the cost of a read
and no embedding.

**4. The published collection is never mutated.** The copy reads it and writes
only into the candidate.

**5. Deleted and renamed paths.** Only paths in the current listing are copied, so
a published path that left the corpus never enters the candidate.

**6. Chunk ids.** Preserved on copy. Agno/Chroma key content by hash, and
regenerating ids would corrupt the candidate.

**7. Publication stays atomic.** The candidate is still built separately and
swapped in.

## Verified against real ChromaDB, not only the fake

The fake can only prove the logic is self-consistent, so the production
`KnowledgeManager` was run against a real `chromadb.PersistentClient` with a
deterministic embedder: three files, 40 chunks.

```
1. cold build: indexed 3 files / 40 chunks, 40 embed calls

2. one file changed: indexed 1, 14 embed calls
   reused 27 chunks, ids all published: True
   max vector difference: 5.960e-08

3. forced rebuild: indexed 3, 41 embed calls

4. copy cancelled after writing its rows
   stranded candidate holds: ['alpha.md', 'beta.md', 'gamma.md']
   after deleting gamma.md and resuming: indexed 1, 8 embed calls
   published paths: ['alpha.md', 'beta.md']

5. candidate survived a failed refresh with 2 claims
   forced rebuild over it: indexed 3, 24 embed calls
   published paths: ['alpha.md', 'beta.md', 'gamma.md']
   checkpoint retired: True
```

Step 4 is the recovery end to end: the stranded candidate holds `gamma.md`, the
next refresh rebuilds and copies again rather than re-embedding (8 calls, not
40), and `gamma.md` does not reach the published index.

**Copied vectors are not bit-identical, and cannot be made so.** Chroma stores a
cosine-space collection as a normalized vector plus its norm in float32 and
reconstructs on read, so a value loses up to one float32 ULP per write. 11 of 27
rows came back bit-identical; the rest differ by 6e-8. Two measurements make that
acceptable:

- The drift **converges** rather than accumulating. Over 100 successive copy hops
  at 1536 dimensions, maximum relative drift stops growing after the second hop
  and stays at 2.6e-7, with cosine similarity to the original at 1 − 2e-15. A
  file untouched for a thousand refreshes is no further from its original vector
  than after two.
- An `l2`-space collection round-trips exactly, so the drift is the cosine
  reconstruction, not the copy.

The nearest *other* chunk in the same corpus sits at cosine distance 0.16 — seven
orders of magnitude above the drift.

## Mutation checks

Every testable production branch was broken deliberately and confirmed caught:
**23 of 23**. Not assumed covered — earlier passes found branches with no test
behind them, which is where several of these tests came from.

| Mutation | Test that failed | Message |
|---|---|---|
| disable reuse entirely | `test_git_pull_that_changes_one_file_only_reindexes_that_file` | `the whole checkout was reindexed for a one-file commit` |
| drop the `IndexingSettings` gate | `test_published_vector_reuse_requires_matching_indexing_settings` | `vectors built under other settings were reused` |
| drop the `status == complete` gate | `test_published_vector_reuse_needs_a_completely_published_index` | `vectors from an unpublished collection were reused` |
| ignore `force_reindex` for the copy | `test_forced_reindex_rebuilds_every_vector_instead_of_copying_it` | `a forced rebuild reused the vectors it was asked to replace` |
| stop discarding a surviving candidate when forced | `test_forced_reindex_discards_a_candidate_that_survived_a_failed_refresh` | `a forced rebuild kept the candidate's claims` |
| drop `limit=` from the copy query | `test_published_vector_copy_is_paged_so_no_query_exceeds_the_store_ceiling` | `assert 12 == 1` |
| drop the managed-path filter | `test_published_vector_reuse_skips_paths_that_left_the_corpus` | `assert ['doc0000.md', ...] == ['doc0001.md', 'doc0002.md']` |
| drop the empty-page write guard | `test_published_vector_reuse_skips_paths_that_left_the_corpus` | `assert [] == ['doc0001.md', 'doc0002.md']` |
| narrow the copy-failure fallback | `test_published_vector_reuse_falls_back_when_the_published_collection_is_gone` | `chromadb.errors.NotFoundError` |
| drop the candidate reset after a failed copy | `test_a_failed_vector_copy_leaves_no_partial_candidate` | `partially copied rows survived the failure` |
| stop marking copied paths verified | `test_published_vector_copy_is_paged_so_no_query_exceeds_the_store_ceiling` | `assert 11 <= 4` |
| corrupt the digest in the recorded signature | `test_refresh_after_publish_reuses_vectors_for_unchanged_files` | `the whole corpus was embedded again` |
| claim paths from the listing instead of from copied rows | `test_published_vector_reuse_never_claims_a_path_the_published_index_cannot_serve` | `assert 0 == 1` |
| regenerate chunk ids on copy | `test_refresh_after_publish_reuses_vectors_for_unchanged_files` | `copied chunks must keep their published ids` |
| write into the published collection instead of the candidate | `test_published_vector_reuse_leaves_the_published_collection_untouched` | copy never terminates, killed by `pytest-timeout` |
| skip saving reused signatures into the checkpoint | `test_refresh_after_publish_reuses_vectors_for_unchanged_files` | `the whole corpus was embedded again` |
| stop rebuilding a candidate holding unclaimed rows | `test_a_cancelled_vector_copy_never_publishes_rows_it_could_not_claim` | `a departed path survived in the published index` |
| condemn a candidate that has claims | `test_an_interruption_after_the_copy_still_resumes_the_work_it_finished` | `assert 4 == (4 - 2)` |
| condemn a candidate whose collection is empty | `test_a_candidate_with_no_vectors_and_no_claims_is_still_resumed` | `the retry ledger was reset` |
| record claims per page instead of once at the end | `test_a_copy_records_its_claims_only_after_every_row_has_landed` | `claims were recorded before every copied row had landed` |
| probe candidate existence after building `Knowledge` | `test_a_candidate_whose_collection_vanished_is_rebuilt_rather_than_resumed` | `claims survived a collection that no longer holds their vectors` |
| let a recoverable write failure replace cancellation | `test_shielded_write_keeps_cancellation_authoritative_when_the_write_fails` | `RuntimeError: recoverable write failed` |
| suppress a failed publish-metadata write after cancellation | `test_cancelled_publish_metadata_save_surfaces_a_failed_write` | `DID NOT RAISE RuntimeError` |

Two conditionals are not in that table, both stated rather than hidden:

- `persisted_state.collection is None` is type narrowing for the copy call, and
  unreachable at runtime as noted above.
- **`asyncio.shield` in `_shielded_write` has no mutation pair, because removing
  it changes nothing observable.** The write goes through `asyncio.to_thread`,
  which is itself uncancellable, so the file lands either way; the shield only
  makes the coroutine wait for it before propagating the cancellation, which
  matters solely if the loop closes and the process exits first. That window
  resisted a deterministic test. It is kept because it closes it, matches the
  module's existing pattern, and costs nothing — but it is untested and saying so
  is better than a test that only looks like it covers it.

## Test fake changes

Both knowledge test fakes grew what the copy needs, so neither silently takes the
exception fallback: stored chunk ids, `include=["embeddings", "documents"]`, and
an `add` that rejects an empty write the way real Chroma does (`Expected
Embeddings to be non-empty list or numpy array, got [] in add.`, verified against
1.5.8). `chroma_get_result` gained `documents` and `embeddings` and now raises
when `include` asks for a field the fake was not given, so a query shape it
cannot serve says so instead of looking like an empty store.

Without this, ~140 tests in `test_knowledge_manager.py` would have exercised the
fallback instead of the copy, and the paging and reset branches would have had no
coverage at all.

## Deliberately not changed

- **No Git diff pre-filter**, for the reasons above.
- **The 500-row page size is not adaptive.** #1707 adapts the *verification*
  query by splitting on failure; a fixed page needs no such machinery because it
  never asks for more than it can be given.

## Defect found and not fixed

Chroma's cosine-space storage perturbs every vector it writes by up to one
float32 ULP, so a vector is never bit-stable across a write even when the value
came straight out of the same store. It converges after two hops and sits seven
orders of magnitude under the nearest-neighbour gap, so it needs no action here —
but anything that ever compares stored vectors for equality should compare with a
tolerance, not with `==`.

## Summary by CodeRabbit

* **New Features**
  * Knowledge-base refreshes reuse vectors for unchanged files to reduce unnecessary re-embedding.
  * Added a force option to perform a full reindex (rebuilds every vector without reuse).
* **Bug Fixes**
  * Improved interruption/cancellation handling so refreshes publish only after durable completion and can safely resume or reconcile.
* **Documentation**
  * Clarified refresh reuse rules and when stored vectors are invalidated (chunking, embedder, corpus filters), plus how explicit reindex differs.
* **Tests**
  * Expanded reuse/resume/forced-reindex coverage, added Git-based partial-reindex validation, and strengthened cancellation and paging/edge-case assertions.

### CODE REVIEW

File: src/mindroom/knowledge/manager.py [hunk 1]
```diff
@@ -567,6 +512,22 @@ def knowledge_source_signature(
...
+    """Return the managed source path and signature one published chunk carries."""
+    source_path = metadata.get(_SOURCE_PATH_KEY)
+    if not isinstance(source_path, str) or source_path not in managed_paths:
+        return None
+    source_mtime_ns = metadata.get(_SOURCE_MTIME_NS_KEY)
+    source_size = metadata.get(_SOURCE_SIZE_KEY)
+    source_digest = metadata.get(_SOURCE_DIGEST_KEY)
+    if not isinstance(source_mtime_ns, int) or not isinstance(source_size, int) or not isinstance(source_digest, str):
+        return None
+    return source_path, (source_mtime_ns, source_size, source_digest)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/knowledge/manager.py]: Fixed in c9722cf36: published rows now reject boolean mtimes/sizes, negative sizes, and empty digests before reuse. Added regression coverage for each malformed signature.

File: src/mindroom/knowledge/manager.py [hunk 2]
```diff
@@ -1755,42 +1933,45 @@ async def _open_candidate_run(self) -> _CandidateRun:
...
-            else:
-                logger.warning(
-                    "Knowledge candidate collection is missing; rebuilding it from scratch",
-                    base_id=self.base_id,
-                    collection=checkpoint.collection,
-                )
-                checkpoint = replace(checkpoint, completed={}, failed={})
-                checkpoint = await asyncio.to_thread(save_candidate_checkpoint, self._base_storage_path, checkpoint)
-                await asyncio.to_thread(self._reset_vector_db, vector_db)
+            opened = await self._resume_candidate_collection(checkpoint, embedder=embedder)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/knowledge/manager.py]: Already fixed on the current head: force_reindex discards any surviving candidate before opening a run and suppresses published-vector reuse. Focused forced-reindex regressions pass, so no additional change was needed.

File: src/mindroom/knowledge/manager.py [hunk 3]
```diff
@@ -1709,7 +1686,208 @@ async def _candidate_paths_missing_vectors(self, run: _CandidateRun, relative_pa
...
+            # copy must cost a full rebuild rather than a failed refresh. The
+            # partially filled candidate is reset so the rebuild starts clean.
+            logger.warning(
+                "Falling back to a full knowledge rebuild after the published index could not be copied",
+                base_id=self.base_id,
+                collection=published_collection,
+                exc_info=True,
+            )
+            await asyncio.to_thread(self._reset_vector_db, candidate_vector_db)
+            return {}
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/knowledge/manager.py:1915]: Fixed in c9722cf36: the published-vector copy now runs as an explicit owned task that is shielded and drained before cancellation can release the refresh lock. The same drain defers repeated cancellation for durable claim writes. Real Task.cancel and blocked-worker regressions cover both races.
